src/docmanager/index.ts

```diff
@@ -0,0 +1,940 @@
+        if (Private.nameProperty.get(widget) === widgetName) {
+          return widget;
+        }
+      }
+    }
+  }
+
+  /**
+   * Save the document contents to disk.
+   *
+   * #### Notes
+   * This will affect the contents of all other widgets
+   * that share the same model as the given widget.
+   */
+  save(widget: Widget): Promise<void> {
+    let id = Private.contextProperty.get(widget);
+    return this._contextManager.save(id);
+  }
+
+  /**
+   * Save a widget to a different file name.
+   *
+   * #### Notes
+   * It is assumed that all other widgets associated with the new path
+   * have been closed and that the path is either not in conflict
+   * or the user has chosen to overwrite the file.
+   * This will affect the contents of all other widgets
+   * that share the same model as the given widget.
+   */
+  saveAs(widget: Widget, path: string): Promise<void> {
+    let id = Private.contextProperty.get(widget);
+    this._contextManager.rename(id, path);
+    return this._contextManager.save(id);
+  }
+
+  /**
+   * Revert the document contents to disk contents.
+   *
+   * #### Notes
+   * This will affect the contents of all other widgets
+   * that share the same model as the given widget.
+   */
+  revert(widget: Widget): Promise<void> {
+    let id = Private.contextProperty.get(widget);
+    return this._contextManager.revert(id);
+  }
+
+  /**
+   * Close the widgets associated with a given path.
+   */
+  closeFile(path: string): void {
+    let ids = this._contextManager.getIdsForPath(path);
+    for (let id of ids) {
+      let widgets: Widget[] = this._widgets[id] || [];
+      for (let w of widgets) {
+        w.close();
+      }
+    }
+  }
+
+  /**
+   * Close all of the open documents.
+   */
+  closeAll(): void {
+    for (let id in this._widgets) {
+      for (let w of this._widgets[id]) {
+        w.close();
+      }
+    }
+  }
+
+  /**
+   * Create a context or reuse an existing one.
+   */
+  private _createContext(path: string, model: IDocumentModel, widgetName: string, contents: IContentsModel): string {
+    let mFactoryEx = this._getModelFactoryEx(widgetName);
+    let id = this._contextManager.findContext(path, mFactoryEx.name);
+    if (id) {
+      return id;
+    } else {
+      return this._contextManager.createNew(path, model, mFactoryEx, contents);
+    }
+  }
+
+  /**
+   * Create a widget from a context and attach it to the parent.
+   */
+  private _createWidget(contextId: string, widgetName: string, parent: Widget, kernel?: IKernelId): void {
+    let wFactoryEx = this._getWidgetFactoryEx(widgetName);
+    if (!(contextId in this._widgets)) {
+      this._widgets[contextId] = [];
+    }
+    this._widgets[contextId].push(parent);
+    let context = this._contextManager.getContext(contextId);
+    let model = this._contextManager.getModel(contextId);
+    // Create the child widget using the factory.
+    let child = wFactoryEx.factory.createNew(model, context, kernel);
```

Comment:
The widgetFactory interface defines `kernel` is a mandatory parameter, but it's optional here. How about changing the widgetFactory.createNew interface to have kernel be optional?